### ISSUE TITLE
Subscription types incompatible with graphql-subscription's PubSub

### ISSUE THREAD

**Issue by sarink** [dotansimha/graphql-code-generator#7197]
The latest version of graphql-code-generator has changed the subscribe function type to `AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;`

However, when using apollo's graphql-subscriptions, the type returned from pubsub-engine is an `AsyncIterator` https://github.com/apollographql/graphql-subscriptions/blob/master/src/pubsub-engine.ts#L8

`AsyncIterable` and `AsyncIterator` types are incompatible.

What's the recommended workaround for this issue?

**Comment by armintan** [comment]
I do habe the same problem ... any recommendations?

**Comment by Thompson-Jordon** [comment]
I also have this same issue. 

`import { SubscriptionResolvers } from '../../generated/graphql'
import { Context } from '../context'

export const NEW_POINT = 'NEW_POINT'

export const Subscription: SubscriptionResolvers<Context> = {
  newPoint: {
    subscribe: (_,__,context) => context.pubsub.asyncIterator(NEW_POINT)
  },
}`

**Comment by mpokrovsky** [comment]
https://github.com/dotansimha/graphql-code-generator/pull/7015#issuecomment-975904577

**Comment by Urigo** [comment]
@mpokrovsky I think graphql-subscriptions 3.0 might be happening soon and I think it includes the changes from @n1ru4l 

Could you check if using the current state of their PR solves this issue? https://github.com/apollographql/graphql-subscriptions/pull/250

**Comment by Mad-Kat** [comment]
In the meantime you could use something like this:
```typescript
subscribe: () => ({
        [Symbol.asyncIterator]: () => pubsub.asyncIterator<{ numberIncremented: number }>('NUMBER_INCREMENTED'),
      }),
```

**Comment by n1ru4l** [comment]
@Mad-Kat This would be a working, but rather an ugly workaround. You should not tailor your code and add overhead for pleasing a library with wrong typings.

I recommend everyone to read through https://github.com/dotansimha/graphql-code-generator/pull/7015 for a full breakdown of the situation and migrate off graphql-subscriptions or use the mentioned patch-package for fixing the typings.

**Comment by maxpain** [comment]
Any updates?

**Comment by n1ru4l** [comment]
I will close this issue as there is nothing wrong within graphql-code-generator.

**Comment by loucadufault** [comment]
Workaround proposed in above https://github.com/dotansimha/graphql-code-generator/issues/7197#issuecomment-1098014584 with `graphql-subscriptions` `withFilter`:

```ts
import { withFilter } from 'graphql-subscriptions';

export const resolvers = {
  Subscription: {
    somethingChanged: {
      subscribe: () => ({
        [Symbol.asyncIterator]: withFilter(
          (_, args) => pubsub.asyncIterator(`${SOMETHING_CHANGED_TOPIC}.${args.relevantId}`),
          (payload, variables: SubscriptionSomethingChangedArgs) => payload.somethingChanged.id === variables.relevantId,
        ),
      }),
    },
  },
}
```

**Comment by skizzo** [comment]
@loucadufault your solution doesn't throw any TypeScript errors indeed, but `variables` is always undefined, so I can't filter by the vars the client defines when adding the subscription. Any ideas?

**Comment by loucadufault** [comment]
@skizzo you are correct. I had noticed before (and downvoted my own comment) but forgot to add a note. I have now amended it.

The only thing I have found to work is to cast as `any`, sadly.

**Comment by skizzo** [comment]
I found out that when using [`graphql-subscriptions-continued`](https://github.com/apollographql/graphql-subscriptions)'s `withFilter` (instead of the `graphql-subscriptions` one), the IDE errors disappear and everything works as it supposed to. Unfortunately, `payload`, `args` and `context` arguments are still not typed, but at least it works without errors.
